ICU-20328 Implement LocaleBuilder

FrankYFTang opened the pull request:
Design Doc: https://goo.gl/Qf12p3

<!--
Thank you for your pull request.
Please see http://site.icu-project.org/processes/contribute for general
information on contributing to ICU.

You will be automatically asked to sign the contributors license before the PR is accepted.
- sign: https://cla-assistant.io/unicode-org/icu
- license: http://www.unicode.org/copyright.html#License
-->

##### Checklist

- [X] Issue filed: https://unicode-org.atlassian.net/browse/ICU-20328
- [X] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [X] Tests included
- [X] Documentation is changed or added

FrankYFTang commented:
ping I fixed both issues @yumaoka  mentioned.

yumaoka commented:
Thanks - looks good for the updates.

There are still test failures and build problem on Windows.

For the test failure like below -

https://travis-ci.org/unicode-org/icu/jobs/479685426
```
    TestForLanguageTag   {
!    uloc_forLanguageTag returned locale [] for input language tag [hant-cmn-cn] - expected: [hant]
!!    uloc_forLanguageTag parsed length of 0 for input language tag [hant-cmn-cn] - expected parsed length: 4
    } ---[2 ERRORS in /tsutil/cloctst/TestForLanguageTag]  
```

Is this a regression issue?

jefgen commented:
Regarding the Windows build failure/problem(s): @FrankYFTang I made a copy of your branch in my fork and made a commit that fixes the build issues.

Link: https://github.com/jefgen/icu/commit/b382946a572b936fdc850dd54ed8395c0e1f5ab0

Feel free to cherry-pick the commit into your branch, or just copy/paste the changes over. :)

FrankYFTang commented:
@markusicu Updated and fixed issues. Please start revewing

FrankYFTang commented:
> Thanks - looks good for the updates.
> 
> There are still test failures and build problem on Windows.
> 
> For the test failure like below -
> 
> https://travis-ci.org/unicode-org/icu/jobs/479685426
> 
> ```
>     TestForLanguageTag   {
> !    uloc_forLanguageTag returned locale [] for input language tag [hant-cmn-cn] - expected: [hant]
> !!    uloc_forLanguageTag parsed length of 0 for input language tag [hant-cmn-cn] - expected parsed length: 4
>     } ---[2 ERRORS in /tsutil/cloctst/TestForLanguageTag]  
> ```
> Is this a regression issue?

Addressed. This is related to 4alpha for language. Per 2019-1-23 ICUTC we still accept it so I changed the code to allow it now.

FrankYFTang commented:
@markusicu ping

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+    locale->setKeywordValue(kAttributeKey, "", error);
+
+    // Clear all Unicode keyword values
+    StringEnumeration* str_enum = locale->createUnicodeKeywords(error);
```
```
    LocalPointer<StringEnumeration> str_enum(locale->createUnicodeKeywords(error));
    if (U_SUCCESS(error)) {
        const char* key;
        while ((key = str_enum->next(nullptr, error)) != nullptr) {
            locale->setUnicodeKeywordValue(key, nullptr, error);
        }
    }
```

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+        status_ = U_ILLEGAL_ARGUMENT_ERROR;
+        return *this;
+    }
+    CharString value_str(value, status_);
```
After this line:
```
if (U_FAILURE(status_)) {
    return *this;
}
```

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+        return *this;
+    }
+    _clearUAttributesAndKeyType(extensions_, status_);
+    if (value.length() > 0) _setUnicodeExtensions(extensions_, value_str, status_);
```
{}

markusicu replied:
value.length() > 0   -->   !value.empty()

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+{
+    if (U_FAILURE(status_)) return *this;
+    if (!ultag_isUnicodeLocaleKey(key.data(), key.length()) ||
+        (type.length() > 0 &&
```
type.length() > 0   -->   !type.empty()

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+    for (int32_t i = 0; i < value_str.length(); i++, p++) {
+        *p = (*p == '_') ? '-' : uprv_tolower(*p);
+    }
+    if (value_str.length() > 0 &&
```
value.length() > 0 --> !value.empty()

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+      status_ = U_ILLEGAL_ARGUMENT_ERROR;
+      return *this;
+    }
+    if (extensions_ == nullptr) extensions_ = new Locale();
```
{}
and check for alloc failure

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+        return *this;
+    }
+    if (extensions_ == nullptr) {
+        extensions_ = new Locale();
```
check for alloc failure

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+    CharString value_str(value, status_);
+    if (U_FAILURE(status_)) return *this;
+    char* p = value_str.data();
+    for (int32_t i = 0; i < value_str.length(); i++, p++) {
```
Please make a helper function that takes a non-const char* and length and does the _- and lowercasing -- and replace each of these loops.

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+        return *this;
+    }
+    if (extensions_ == nullptr) {
+        extensions_ = new Locale();
```
check for alloc failure

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+    }
+    if (extensions_ == nullptr) {
+        extensions_ = new Locale();
+        extensions_->setKeywordValue(kAttributeKey, value_str.data(), status_);
```
This line & next need to go outside the if. Make sure you have a test case that would fail here.

FrankYFTang replied:
no, this should be inside the == nullptr. If there were NO extension, we just need to allocate one and set one value without any other processing

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+
+    CharString attributes;
+    CharStringByteSink sink(&attributes);
+    UErrorCode get_error = U_ZERO_ERROR;
```
I think you should use status_ here.

If there's a reason to have a separate UErrorCode here, please call it localStatus or localErrorCode or...
We use camelCase like in Java.

FrankYFTang replied:
No I cannot use status_. This is checking we have attribute or not. If it fail to get attribute, nothing wrong here and I cannot set the status to error.

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+    CharString attributes;
+    CharStringByteSink sink(&attributes);
+    UErrorCode get_error = U_ZERO_ERROR;
+    extensions_->getKeywordValue(kAttributeKey, sink, get_error);
```
after this, if you do keep a separate UErrorCode, check for failure

FrankYFTang replied:
Actually, I don't need to. If there are no such attribute, then it will fail, but then the attributes will be empty, and therefore the while loop will just be skipped (since start == end) and therefore the whole thing will work in the end. But I could add a failure check (which mean there were no attribute in the extension so far, not really a real failure) and duplicate the code and return earlier.

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+    }
+
+    const char* start = attributes.data();
+    const char* end = attributes.data() + attributes.length();
```
Optional: We always call an exclusive-end a "limit".

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+    extensions_->getKeywordValue(kAttributeKey, sink, get_error);
+    p = attributes.data();
+    for (int32_t i = 0; i <  attributes.length(); i++, p++) {
+        *p = (*p == '_' || *p == '-') ? '\0' : uprv_tolower(*p);
```
Please add a comment about replacing separators with NUL.

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+        }
+        if (!new_attributes.isEmpty()) new_attributes.append('_', status_);
+        new_attributes.append(start, status_);
+        start += strlen(start) + 1;
```
uprv_strlen

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+        return *this;
+    }
+    if (extensions_ == nullptr) return *this;
+    UErrorCode get_error = U_ZERO_ERROR;
```
localStatus or localErrorCode or...

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+    // get failure, just return
+    if (U_FAILURE(get_error)) return *this;
+    // Do not have any attributes, just return.
+    if (attributes.length() == 0) return *this;
```
.isEmpty()

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+
+    p = attributes.data();
+    for (int32_t i = 0; i < attributes.length(); i++, p++) {
+        *p = (*p == '_' || *p == '-') ? '\0' : uprv_tolower(*p);
```
comment about NUL

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+
+Locale LocaleBuilder::build(UErrorCode& error)
+{
+    if (U_SUCCESS(error)) {
```
```
    if (U_FAILURE(status or errorCode)) {
        Locale bogus;
        bogus.setToBogus();
        return bogus;
    }
    if (U_FAILURE(status_)) {
        errorCode = status_;
        Locale bogus;  // could make a makeBogusLocale() helper function
        bogus.setToBogus();
        return bogus;
    }
    ... good case ...
```

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+        if (uprv_strlen(script_) > 0) {
+            lang_script.append('-', error).append(StringPiece(script_), error);
+        }
+        if (U_SUCCESS(error)) {
```
less nesting...
```
if (U_FAILURE(status)) {
    return makeBogusLocale();
}
```

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+{
+    if (U_SUCCESS(error)) {
+        error = status_;
+        CharString lang_script(language_, error);
```
interesting but a little confusing

I would either use a char langScript[14] and just uprv_strcat() or similar,

or else make a single CharString with all of the non-empty lang-script-region-variant and use the single-string Locale constructor.

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+
+void LocaleBuilderTest::TestSetLanguageWellFormed() {
+    // http://www.unicode.org/reports/tr35/tr35.html#unicode_language_subtag
+    // unicode_language_subtag = alpha{2,3} | alpha{5,8};
```
except we decided to also support alpha{4} in ICU, right?

Please also test that.

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+void LocaleBuilderTest::TestSetLanguageWellFormed() {
+    // http://www.unicode.org/reports/tr35/tr35.html#unicode_language_subtag
+    // unicode_language_subtag = alpha{2,3} | alpha{5,8};
+    static const char* well_formed_language[] = {
```
should be plural

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+        "IADSFJKR",
+        "kkDSFJkR",
+    };
+    for (int i = 0; i < UPRV_LENGTHOF(well_formed_language); i++) {
```
Optional: for (const char* lang : well_formed_languages) {

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+void LocaleBuilderTest::TestSetVariantWellFormed() {
+    // http://www.unicode.org/reports/tr35/tr35.html#unicode_variant_subtag
+    // (sep unicode_variant_subtag)*
+    // unicode_variant_subt a= (alphanum{5,8} | digit alphanum{3}) ;
```
typo   unicode_variant_subt a

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+        bld.setUnicodeLocaleKeyword("ab", ill_formed[i]);
+        Locale loc = bld.build(status);
+        if (status != U_ILLEGAL_ARGUMENT_ERROR) {
+            errln("setUnicodeLocaleKeyword(\"%s\") should fail but has no Error\n",
```
mention key `\"ab\"` like you mentioned the type in the previous function

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+void LocaleBuilderTest::TestAddRemoveUnicodeLocaleAttributeWellFormed() {
+    // http://www.unicode.org/reports/tr35/tr35.html#unicode_locale_extensions
+    // attribute = alphanum{3,8} ;
+    static const char* well_formed_attribute[] = {
```
Please fix all variables to camelCase.

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+
+    bld.setExtension('U', "ca-islamic");
+    Verify(bld, "zh-u-ca-islamic",
+           "setExtension('t', \"zh-u-ca-islamic\") should success but got Error: %s\n");
```
U not t

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+
+    bld.setRegion("TW");
+    Verify(bld, "zh-TW-u-co-pinyin",
+           "setRegion(\"MX\") should success but got Error: %s\n");
```
TW not MX

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+    };
+    char ch = 'a';
+    for (int i = 0; i < UPRV_LENGTHOF(well_formed); i++, ch++) {
+        if (ch > 'z') ch = 'a';
```
This will fail with EBCDIC where a-z are three ranges rather than one (same for A-Z).

I suggest you use a string constant "abcde...wxyz" and when you get to index 26 or pick up the terminating NUL you reset the index to 0.

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+        LocaleBuilder bld;
+        bld.setExtension(ch, well_formed[ch % UPRV_LENGTHOF(well_formed)]);
+        Locale loc = bld.build(status);
+        if (('a' <= ch && ch <= 'z') || ('A' <= ch && ch <= 'Z') || ('0' <= ch && ch <= '9')) {
```
Won't work with EBCDIC. I suggest you use uprv_isASCIILetter(c) for the letters.

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+    };
+
+    // Test key with the whole the ASCII range.
+    for (char ch = 0x20; ch < 0x7F; ch++) {
```
In EBCDIC, this would be 0x40..0xFF.
https://en.wikipedia.org/wiki/EBCDIC_037

I suggest you use a string literal with all letters and digits, and some other stuff, and iterate over that.

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+    };
+    char ch = 'a';
+    for (int i = 0; i < UPRV_LENGTHOF(ill_formed); i++, ch++) {
+        if (ch > 'z') ch = 'a';
```
EBCDIC... see above

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+    Locale l1 = Locale::forLanguageTag("en-US-u-va-posix", status);
+    if (U_FAILURE(status) || l1.isBogus()) {
+        errln("build should success but got Error: %s\n", u_errorName(status));
+    }
```
Test that we should get the same as Locale("en-US-POSIX")? (Assuming we handle that?)

FrankYFTang commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+static void setField(StringPiece input, char* dest, UErrorCode& error,
+                     UBool (*test)(const char*, int32_t)) {
+    if (U_FAILURE(error)) return;
+    if (input.length() == 0) {
```
error: ‘class icu_63::StringPiece’ has no member named ‘isEmpty’; did you mean ‘empty’?
     if (input.isEmpty()) {
               ^~~~~~~

markusicu replied:
yes, sorry -- empty() or isEmpty() depending on the class. looks like you did all of these.

FrankYFTang commented on icu4c/source/common/uloc_tag.cpp:
```diff
@@ -488,19 +497,47 @@ _isVariantSubtag(const char* s, int32_t len) {
+}
+
 static UBool
 _isPrivateuseVariantSubtag(const char* s, int32_t len) {
```
This is something not in the spec, it is where it has "lvariant" 
static const char PRIVUSE_VARIANT_PREFIX[] = "lvariant";
Not sure what it is.

markusicu replied:
Please add a comment mentioning the ICU-specific "lvariant" handling.

FrankYFTang commented on icu4c/source/common/uloc_tag.cpp:
```diff
@@ -528,101 +565,219 @@ _isExtensionSubtag(const char* s, int32_t len) {
+}
+
+static UBool
+_isPrivateuseValueSubtag(const char* s, int32_t len) {
```
The other one is something I am not sure it should be and I think it is better not to touch that or combine these two.

FrankYFTang commented on icu4c/source/common/uloc_tag.cpp:
```diff
@@ -528,101 +565,219 @@ _isExtensionSubtag(const char* s, int32_t len) {
+            // no break nor return here.
+        case kGotVariant:
+            if (_isVariantSubtag(s, len)) {
+                *state = kGotVariant;
```
????

FrankYFTang commented on icu4c/source/common/uloc_tag.cpp:
```diff
@@ -528,101 +565,219 @@ _isExtensionSubtag(const char* s, int32_t len) {
+                return TRUE;
+            }
+            if (_isTValue(s, len)) {
+                *state = kGotTValue;
```
????? why? I am following the spec. Notice the "+"
tvalue | = (sep alphanum{3,8})+ ;

http://unicode.org/reports/tr35/#Unicode_locale_identifier

FrankYFTang commented on icu4c/source/common/uloc_tag.cpp:
```diff
@@ -528,101 +565,219 @@ _isExtensionSubtag(const char* s, int32_t len) {
+                *state = kGotKey;
+                return TRUE;
+            }
+            return FALSE;
```
good catch. I missed that.

FrankYFTang commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+        bld.clear();
+        while (true) {
+            method = TESTCASE[tidx][i++];
+            if (strcmp("L", method) == 0) {
```
I try to keep the code close to Java.

FrankYFTang commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -0,0 +1,1607 @@
+
+    bld.setUnicodeLocaleKeyword("co", "");
+    Verify(bld, "zh-Hans-HK-revised",
+           "setUnicodeLocaleKeyword('nu', '') should success but got Error: %s\n");
```
That depend on the fix in the icu::Locale, which have not yet being fixed yet. I will do that in a separate PR.

FrankYFTang commented:
@markusicu  PTAL

nciric commented on icu4c/source/common/unicode/localebuilder.h:
```diff
@@ -0,0 +1,288 @@
+ *
+ * <p>LocaleBuilder track errors in an internal UErrorCode. For all methods
+ * setters except, setLanguageTag, setLocale, clear and clearExtension,
+ * LocaleBuilder will return immediately if the internal UErrorCode is in
```
I would rewrite this into:

For all setters, except setLanguageTag, setLocale, clear and clearExtension, LocaleBuilder...

markusicu replied:
Oh -- this should not be the case for clearExtension(). The exceptions should be only clear() and the assignment-like methods. Please check the code.

FrankYFTang replied:
sorry. I got confused about the clearExtension and put in the wrong comment here. The code is correct. My newly added comment was wrong.

FrankYFTang replied:
comment corrected. PTAL

markusicu commented on icu4c/source/common/unicode/localebuilder.h:
```diff
@@ -44,232 +45,240 @@ class CharString;
+     * Sets the Unicode locale keyword type for the given key. If the type
+     * StringPiece is constructed with a nullptr, the keyword is removed.
+     * If the type is the empty string, the keyword is set without type subtags.
+     * Otherwise, the or else the build() method will later report an
```
Otherwise, the key and type must be well-formed, or else ...

markusicu commented on icu4c/source/common/uloc_tag.cpp:
```diff
@@ -668,54 +668,52 @@ _isTransformedExtensionSubtag(int32_t* state, const char* s, int32_t len)
-            // no break nor return here.
+            U_FALLTHROUGH;
         case kGotVariant:
             if (_isVariantSubtag(s, len)) {
```
You do need
```
state = kGotVariant;
```
after this line because of the fall-throughs.

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -0,0 +1,395 @@
+        return *this;
+    }
+    CharString* new_variant = new CharString(variant, status_);
+    if (new_variant == nullptr) {
```
not done

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -1,4 +1,4 @@
+// © 2019 and later: Unicode, Inc. and others.
 // License & terms of use: http://www.unicode.org/copyright.html
 
 #include <memory>
```
Did you include this for std::unique_ptr? Still needed?

markusicu commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -870,14 +877,14 @@ void LocaleBuilderTest::TestSetUnicodeLocaleKeywordIllFormedValue() {
         if (status != U_ILLEGAL_ARGUMENT_ERROR) {
-            errln("setUnicodeLocaleKeyword(\"%s\") should fail but has no Error\n",
-                  ill_formed[i]);
+            errln("setUnicodeLocaleKeyword(\"abc\", \"%s\") should fail but has no Error\n",
```
ab not abc

nciric commented on icu4c/source/common/unicode/localebuilder.h:
```diff
@@ -0,0 +1,287 @@
+ *
+ * <p>LocaleBuilder track errors in an internal UErrorCode. For all setters,
+ * except setLanguageTag and setLocale, LocaleBuilder will return immediately
+ * if the internal UErrorCode is in error state. clear will reset the error code.
```
To start sentence with capital letter:

To reset internal state and error code, call clear method.

or 

Call clear to reset internal state and error code.

FrankYFTang commented on icu4c/source/test/intltest/localebuildertest.cpp:
```diff
@@ -1597,11 +1614,17 @@ void LocaleBuilderTest::TestPosixCases() {
+    Locale posix("en-US-POSIX");
+    if (posix != l2) {
+        errln("The result locale should be the set as %s but got %s\n",
+              posix.toLanguageTag<std::string>(status).c_str(),
```
it is test on "if (posix != l2)" the toLanguageTag is just print out the error message. I change it to getName() now.

markusicu replied:
Sorry, I see now that you did compare Locale objects.

FrankYFTang commented:
Thanks for your review feedbacks. PTAL

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -167,14 +166,17 @@ _copyExtensions(const Locale& from, Locale* to, bool validate, UErrorCode& error
+            return;
         }
         to->setKeywordValue(key, value.data(), errorCode);
+        if (U_FAILURE(errorCode)) { return; }
```
Optional: Ok, but should not be necessary -- iter->next() should return nullptr if failure.

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -167,14 +166,17 @@ _copyExtensions(const Locale& from, Locale* to, bool validate, UErrorCode& error
         CharString value;
         CharStringByteSink sink(&value);
         from.getKeywordValue(key, sink, errorCode);
+        if (U_FAILURE(errorCode)) { return; }
```
Thanks for adding this failure check.
Just to explain, if getKeywordValue() fails, then the value is not usable, and we should not use value.data(), so we have to check here.

FrankYFTang commented:
PTAL

markusicu commented on icu4c/source/common/localebuilder.cpp:
```diff
@@ -398,23 +402,42 @@ Locale makeBogusLocale() {
+            return makeBogusLocale();
+        }
     }
     Locale product(locale_str.data());
```
ping
